types/wlr_data_device_manager.c

```diff
@@ -0,0 +1,185 @@
+	// struct wlr_data_device *device = handle->wlr_seat->data_device;
+	// struct wlr_data_source *src = wl_resource_get_user_data(src_res);
+	// struct wlr_surface *origin = wl_resource_get_user_data(origin_res);
+	// struct wlr_surface *icon;
+	// if (icon_res)
+	// 	icon = wl_resource_get_user_data(icon_res);
+	// wlr_seat_start_drag(serial, device->seat, src,
+	// 	origin, icon); // will set surface roles and emit signal for user
+}
+
+static void wl_cb_data_device_set_selection(struct wl_client *client,
+		struct wl_resource *res, struct wl_resource *src_res,
+		uint32_t serial) {
+	// TODO: serial validation
+	struct wlr_seat_handle *handle = wl_resource_get_user_data(res);
+	struct wlr_data_device *device = handle->wlr_seat->data_device;
+	struct wlr_data_source *src = wl_resource_get_user_data(src_res);
+	wlr_data_device_set_selection(device, src);
+}
+
+static struct wl_data_device_interface data_device_impl = {
+	.start_drag = wl_cb_data_device_start_drag,
+	.set_selection = wl_cb_data_device_set_selection,
+	.release = resource_destroy
+};
+
+static void data_device_selection_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_data_device *device = wl_container_of(listener, device, selection_destroyed);
+	assert(data == device->selection);
+	device->selection = NULL; // make sure no cancel is sent
+	wlr_data_device_set_selection(device, NULL);
+}
+
+static struct wlr_data_device *seat_ensure_data_device(struct wlr_data_device_manager *manager,
+		struct wlr_seat *seat) {
+	if (seat->data_device) {
+		return seat->data_device;
+	}
+
+	if (!(seat->data_device = calloc(1, sizeof(*seat->data_device)))) {
+		wlr_log(L_ERROR, "Failed to allocate wlr_data_device");
+		return NULL;
+	}
+
+	seat->data_device->seat = seat;
+	wl_signal_init(&seat->data_device->events.selection_change);
+	seat->data_device->selection_destroyed.notify = data_device_selection_destroy;
+	return seat->data_device;
+}
+
+static void data_device_destroy(struct wl_resource *res) {
+	struct wlr_seat_handle *handle = wl_resource_get_user_data(res);
+	handle->data_device = NULL;
+}
+
+static void data_device_manager_create_data_source(struct wl_client *client,
+		struct wl_resource *res, uint32_t id) {
+	uint32_t version = wl_resource_get_version(res);
+	if (!wlr_wl_data_source_create(client, version, id)) {
+		wlr_log(L_ERROR, "Failed to create wlr_wl_data_source");
+		wl_resource_post_no_memory(res);
+		return;
+	}
+}
+
+static void data_device_manager_get_data_device(struct wl_client *client,
+		struct wl_resource *res, uint32_t id, struct wl_resource *seat_res) {
+	struct wlr_data_device_manager *manager = wl_resource_get_user_data(res);
+	struct wlr_seat_handle *seat_handle = wl_resource_get_user_data(seat_res);
+	struct wlr_data_device *device;
+	if (!(device = seat_ensure_data_device(manager, seat_handle->wlr_seat))) {
+		wl_resource_post_no_memory(res);
+		return;
+	}
+
+	if (seat_handle->data_device) {
```

Comment:
I don't even think it would make sense it this case. As tried to describe above data_device and seat are really closely connected.